# Data Cleaning with Pandas

In [85]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [86]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')

/Users/teo/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [87]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

In [88]:
sales_df.iloc[:,1].head()

0    138860
1    664885
2    423943
3    403700
4    951120
Name: Major, dtype: object

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [89]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [90]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.5+ MB


In [91]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/Users/teo/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [92]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [93]:
bldg_df.head().T

0                            1  \
Major                                    4300                         4610   
Minor                                     167                          399   
BldgNbr                                     1                            1   
NbrLivingUnits                              1                            1   
Address             15223   40TH AVE S  98188  4431   FERNCROFT RD   98040   
BuildingNumber                          15223                        4431    
Fraction                                                                     
DirectionPrefix                                                              
StreetName          40TH                         FERNCROFT                   
StreetType                             AVE                          RD       
DirectionSuffix                            S                                 
ZipCode                                 98188                        98040   
Stories                                     1                            2   
BldgGrade                                   6                            9   
BldgGradeVar                                0                            0   
SqFt1stFloor                              990                          980   
SqFtHalfFloor                               0                            0   
SqFt2ndFloor                                0                         1140   
SqFtUpperFloor                              0                            0   
SqFtUnfinFull                               0                            0   
SqFtUnfinHalf                               0                            0   
SqFtTotLiving                             990                         2950   
SqFtTotBasement                           990                          940   
SqFtFinBasement                             0                          830   
FinBasementGrade                            0                            9   
SqFtGarageBasement                          0                            0   
SqFtGarageAttached                          0                          480   
DaylightBasement                                                         Y   
SqFtOpenPorch                              80                            0   
SqFtEnclosedPorch                           0                            0   
SqFtDeck                                    0                          600   
HeatSystem                                  6                            5   
HeatSource                                  1                            2   
BrickStone                                  0                            0   
ViewUtilization                                                          N   
Bedrooms                                    2                            3   
BathHalfCount                               0                            1   
Bath3qtrCount                               0                            0   
BathFullCount                               1                            2   
FpSingleStory                               0                            0   
FpMultiStory                                0                            0   
FpFreestanding                              0                            2   
FpAdditional                                0                            0   
YrBuilt                                  1922                         1978   
YrRenovated                                 0                            0   
PcntComplete                                0                            0   
Obsolescence                                0                            0   
PcntNetCondition                            0                            0   
Condition                                   3                            4   
AddnlCost                                   0                            0   

                                              2                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [94]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [95]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511412 entries, 0 to 511411
Data columns (total 4 columns):
Major            511412 non-null int64
Minor            511412 non-null int64
SqFtTotLiving    511412 non-null int64
ZipCode          468372 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [96]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [97]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'])

ValueError: Unable to parse string "      " at position 935519

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [ ]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [98]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [99]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.5+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [100]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [101]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.5+ MB


Now, let's try our join again.

In [102]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [103]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [104]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1437522 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1437522 non-null float64
Minor            1437522 non-null float64
DocumentDate     1437522 non-null object
SalePrice        1437522 non-null int64
SqFtTotLiving    1437522 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 76.8+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [105]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
99   858140.0  376.0   05/22/2012          0            900     NaN
100  858140.0  376.0   11/28/2017          0            900     NaN
152  720319.0  520.0   11/20/2013     699950           2840     NaN
153  720319.0  520.0   09/16/2013          0           2840     NaN
163  894677.0  240.0   12/21/2016     818161           2450     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [106]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [107]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1322148 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1322148 non-null float64
Minor            1322148 non-null float64
DocumentDate     1322148 non-null object
SalePrice        1322148 non-null int64
SqFtTotLiving    1322148 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 70.6+ MB


In [108]:
sales_data = sales_data[sales_data["SalePrice"]!=0]

In [109]:
sales_data = sales_data[sales_data["SqFtTotLiving"] > 0]

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [110]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode'].head()

10168    98042-3001
10169    98042-3001
10170    98042-3001
10171    98042-3001
16528    98199-3014
Name: ZipCode, dtype: object

In [114]:
sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode'] = sales_data['ZipCode'].str.slice(stop=5)

In [113]:
sales_data['ZipCode'].str.slice(stop=5).unique()

array(['98002', '98092', '98008', '98058', '98038', '98031', '98188',
       '98051', '98001', '98108', '98198', '98115', '98118', '98072',
       '98117', nan, '98039', '98155', '98075', '98003', '98103', '98022',
       '98042', '98040', '98133', '98105', '98056', '98102', '98053',
       '98168', '98027', '98011', '98074', '98146', '98024', '98029',
       '98006', '98005', '98028', '98034', '98144', '98030', '98177',
       '98166', '98065', '98112', '98116', '98010', '98199', '98032',
       '98106', '98059', '98070', '98045', '98136', '98125', '98023',
       '98033', '98077', '98109', '98055', '98178', '98052', '98122',
       '98014', '98004', '98119', '98107', '98126', '98007', '98019',
       '98224', '98148', '98047', '98288', '98050', '98354', '98068',
       '98057', '98302', '98083', '98121', '98043', '98113', '98134',
       '98026', '89118', '98104', 'A', '98000', '98204', '98035', '98422',
       '98132', 'B', ' ', '95059', '98097', '8126', '28028', '98189',
       '98

### 3. Add a column for PricePerSqFt



In [115]:
sales_data["PricePerSqFt"] = sales_data["SalePrice"] / sales_data["SqFtTotLiving"]

In [116]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0  138860.0  110.0   08/21/2014     245000           1490   98002   
1  138860.0  110.0   06/12/1989     109300           1490   98002   
2  138860.0  110.0   01/16/2005      14684           1490   98002   
6  423943.0   50.0   07/15/1999      96000            960   98092   
7  423943.0   50.0   01/08/2001     127500            960   98092   

   PricePerSqFt  
0    164.429530  
1     73.355705  
2      9.855034  
6    100.000000  
7    132.812500

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [117]:
def is_in2019(x):
    if (x.endswith("2019")):
        return True
    else:
        return False
    
sales_2019 = sales_data.loc[sales_data['DocumentDate'].apply(is_in2019) == True, :]

In [118]:
sales_2019.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
605   214090.0  110.0   03/06/2019     390000           1780   98042   
651   367890.0   30.0   03/13/2019     923000           2550   98144   
868   350160.0  125.0   02/06/2019     935000           2460   98117   
1063  882790.0  870.0   02/15/2019     890000           2380   98105   
1341   30200.0  645.0   03/04/2019     380000           1750   98001   

      PricePerSqFt  
605     219.101124  
651     361.960784  
868     380.081301  
1063    373.949580  
1341    217.142857

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [131]:
def is_inSeattle(x):
    seattle_zips = ['98101', '98102', '98103', '98104',
                    '98105', '98106', '98107', '98108', 
                    '98109', '98110', '98112', '98115', 
                    '98116', '98117', '98118', '98119',
                    '98121', '98122', '98125', '98126',
                    '98131', '98133', '98134', '98136',
                    '98144', '98146', '98148', '98154',
                    '98155', '98158', '98161', '98164',
                    '98166', '98168', '98174', '98177',
                    '98178', '98188', '98198', '98199']
    
    seattle_zips = [98101, 98102, 98103, 98104, 98105, 98106, 98107, 98108, 98109, 98112, 98115, 98116, 98117, 98118, 98119, 98121, 98122, 98125, 98126, 98133, 98134, 98136, 98144, 98146, 98154, 98164, 98174, 98177, 98178, 98195, 98199]
    return x in seattle_zips
    
sales_in_seattle = sales_2019.loc[sales_2019['ZipCode'].apply(is_inSeattle) == True, :]

In [132]:
sales_in_seattle['ZipCode'].unique()

array([98136.0, 98199.0, 98118.0, 98146.0, 98117.0, 98115.0, 98105.0,
       98103.0, 98122.0, 98106.0, 98112.0, 98116.0, 98133.0, 98144.0,
       98126.0, 98109.0, 98125.0, 98108.0, 98177.0, 98102.0, 98107.0,
       98134.0, 98119.0], dtype=object)

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [136]:
sales_in_seattle['PricePerSqFt'].mean()

515.5888848281156

In [134]:
sales_in_seattle.sort_values('PricePerSqFt')

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
1073090  891050.0    23.0   03/28/2019     620000           3620   98177   
1083249  322620.0   150.0   02/14/2019     550000           2820   98118   
423138   713530.0    65.0   01/16/2019     300000           1450   98118   
911906   892510.0   435.0   01/25/2019     700000           3370   98115   
1318906  337190.0     8.0   03/15/2019     690490           2840   98119   
903058   258680.0    45.0   01/28/2019     445000           1760   98146   
1106933  660006.0    30.0   01/25/2019     600000           2330   98146   
990553   322404.0  9095.0   01/25/2019     340000           1300   98108   
52359    324160.0    27.0   02/26/2019     630000           2390   98118   
257900   982870.0  1140.0   01/15/2019     415000           1550   98112   
530620   983420.0   175.0   02/21/2019     450000           1670   98144   
772030   983420.0   105.0   02/22/2019     630000           2270   98144   
87533    444380.0  1415.0   01/25/2019     699000           2450   98117   
396206   891050.0   215.0   03/07/2019     525000           1760   98133   
23420    431820.0   320.0   02/27/2019     617000           2050   98136   
550513   431570.0   495.0   02/27/2019     895000           2690   98136   
766786   957780.0   325.0   01/31/2019    1412000           4150   98126   
1126558  327860.0  1910.0   03/04/2019     560000           1570   98126   
657008   343850.0   725.0   02/14/2019     530000           1470   98106   
219937   738750.0     5.0   01/18/2019     409000           1110   98106   
1078302  322604.0  9466.0   01/29/2019     695500           1880   98115   
618116   891100.0    77.0   02/18/2019     621000           1640   98133   
1257939  539560.0     5.0   01/09/2019     660000           1740   98134   
429541   327480.0   220.0   03/07/2019     645000           1700   98144   
658630   891100.0   290.0   01/25/2019     593000           1550   98133   
1361321  381240.0   792.0   03/08/2019     495000           1290   98118   
1305204  535020.0   880.0   03/07/2019     450000           1170   98122   
1365347  984230.0   580.0   03/01/2019     608220           1570   98126   
1333161  343850.0   337.0   01/02/2019     547000           1410   98106   
1125397  508140.0   690.0   01/08/2019     660000           1700   98115   
...           ...     ...          ...        ...            ...     ...   
1012863  982870.0  1675.0   02/21/2019     800000           1640   98112   
575273   501600.0  1050.0   01/10/2019     855000           1740   98112   
479552   504230.0   105.0   02/20/2019    1262500           2560   98199   
1361412  721740.0   249.0   02/15/2019     745000           1480   98122   
1191503  983420.0   660.0   02/22/2019     780000           1510   98144   
186919   982270.0   375.0   03/08/2019    1200000           2320   98105   
395038   892610.0   485.0   03/19/2019    1312000           2520   98115   
203528   341360.0   120.0   02/08/2019     775000           1460   98117   
694908   507140.0   685.0   03/20/2019     755000           1420   98115   
528032   503530.0  1103.0   01/23/2019    1110000           2060   98199   
1103768  812210.0   405.0   01/29/2019     680000           1260   98126   
1311318  661300.0   720.0   01/03/2019    2045000           3470   98105   
798491   335950.0   715.0   03/08/2019     619000           1020   98115   
712710   379700.0  1705.0   03/13/2019     767000           1250   98103   
1072156  893410.0    10.0   03/08/2019     760000           1230   98115   
1189264  202504.0  9089.0   01/29/2019    1050000           1690   98102   
580994   323535.0    40.0   01/21/2019    1076800           1700   98133   
311451   982870.0  2060.0   02/13/2019     620000            950   98122   
354209   982870.0    90.0   02/21/2019     889000           1300   98122   
1150043  661300.0  1360.0   02/13/2019    1234000           1660   98105   
1341365  504130.0    95.0   03/04/2019     701000           